### DISCUSSION TITLE
Enabling User object

### DISCUSSION THREAD

**Discussion by MikeStall** [microsoft/Power-Fx#2318] [General]
Power Fx supports a standard `User` object for describing the current user. 
In Power Apps, this is the User() function, but that will be migrated to the User object too. 

The User object has a curated well-known set of properties with strong semantics.  The properties are:

| Property | Type | Details | Example| 
| --- | --- | ---  |  --- |
`FullName` | String | The user's full name |  "John Smith"
`Email` | String | The user's email |  "john@contoso.com"
`DataverseUserId` | guid | The primary key into the Dataverse User's table  |   `LookUp([@Users],  ThisRecord.User = [@User].DataverseUserId).'Primary Email'`
`TeamsMemberId` | String | Id used for Teams  |  
`EntraObjectId` | guid | Guid for user's principle Entra (formerly Azure AD) |   

A host provides a derived instance of the [Microsoft.PowerFx.UserInfo](https://learn.microsoft.com/en-us/dotnet/api/microsoft.powerfx.userinfo?view=powerfx-sdk-latest) class. This describes which properties are supported and has virtual getters to allow fetching the properties. The getters are virtual and async to allow lazily fetching values which may require a network call.    `BasicUserInfo` is a derived class that makes it easy to implement `UserInfo`.

For a host to enable User object, 
1. determine which properties the host will support. This can be set on the symbol APIs,
Call `SymbolTable.AddUserInfoObject(..)` on symbols and provide a string array of the supported keys. 

3. at runtime, passes in the corresponding values via a UserInfo object. 
`Call RuntimeConfig.SetUserInfo(userInfo)`

References:
1. UserInfo:" https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Core/Public/Config/UserInfo.cs
2. Some examples:  https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/UserInfoTests.cs

**Comment by tchinnin** [comment]
Hi @MikeStall 
Isn't it more efficient to use the 3rd attribute of LookUp() function to retrieve a single property such as 'Primary Email' ? 

So using 
`LookUp([@Users],  ThisRecord.User = [@User].DataverseUserId,'Primary Email')`

Instead of 
`LookUp([@Users],  ThisRecord.User = [@User].DataverseUserId).'Primary Email'`

cheers

**Reply by MikeStall** [reply]
Both forms should translate to the same underlying runtime operations.  
Power Fx does extensive optimizations (mostly related to delegation) about which fields are accessed, etc.

**Comment by rnwood** [comment]
Will this transition fix the current oops that in Canvas that according to the docs, the Email property actually returns the UPN? If not, this seems like the perfect time.
